### PR TITLE
fix --record to not fail a successful patch

### PR DESCRIPTION
**Pull request by deads2k**
`--record` in the `kubectl patch` command could actually cause the command the fail even if the patch took.  It would also blindly stomp values that conflicted with other in-flight updates.  

This changes to an update that won't allow it to update the hint when its not actually the last thing to touch it.  

@kubernetes/kubectl @janetkuo

### CODE REVIEW

File: hack/test-cmd.sh [hunk 1]
```diff
@@ -438,9 +438,11 @@ runTests() {
 
   ## Patch pod can change image
   # Command
-  kubectl patch "${kube_flags[@]}" pod valid-pod -p='{"spec":{"containers":[{"name": "kubernetes-serve-hostname", "image": "nginx"}]}}'
+  kubectl patch "${kube_flags[@]}" pod valid-pod --record -p='{"spec":{"containers":[{"name": "kubernetes-serve-hostname", "image": "nginx"}]}}'
   # Post-condition: valid-pod POD has image nginx
   kube::test::get_object_assert pods "{{range.items}}{{$image_field}}:{{end}}" 'nginx:'
+  # Post-condition: valid-pod has the record annotation
+  kube::test::get_object_assert pods "{{range.items}}{{$annotations_field}}:{{end}}" '.*kubernetes.io/change-cause.*'
```
Comment (janetkuo) [inline hack/test-cmd.sh]: Create a variable `change_cause_annotation='.*kubernetes.io/change-cause.*'` so that we can reuse it?

Comment (deads2k) [inline hack/test-cmd.sh]: > Create a variable change_cause_annotation='.*kubernetes.io/change-cause.*' so that we can reuse it?

done

**Comment by deads2k** [comment]
@kargakis ptal

File: pkg/kubectl/cmd/patch.go [hunk 2]
```diff
@@ -139,18 +139,17 @@ func RunPatch(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []stri
...
 	}
 	if cmdutil.ShouldRecord(cmd, info) {
-		patchBytes, err = cmdutil.ChangeResourcePatch(info, f.Command())
-		if err != nil {
-			return err
-		}
-		_, err = helper.Patch(namespace, name, api.StrategicMergePatchType, patchBytes)
-		if err != nil {
-			return err
+		if err := cmdutil.RecordChangeCause(patchedObject, f.Command()); err == nil {
```
Comment (janetkuo) [inline pkg/kubectl/cmd/patch.go]: How about 
```go
if ...; err != nil {
  break
}
// Then do the replace here
```

**Comment by janetkuo** [comment]
One suggestion; LGTM otherwise

File: pkg/kubectl/cmd/patch.go (hunk 2 above)
Comment (deads2k) [inline pkg/kubectl/cmd/patch.go]: This isn't in a loop or a separate method to do a `return` in.  I think this is the tightest representation.

Comment (janetkuo) [inline pkg/kubectl/cmd/patch.go]: Oh, you're right. Please ignore this.

**Pull request merged**